iotdb-core/datanode/src/main/java/org/apache/iotdb/db/service/DataNodeServerCommandLine.java

```diff
@@ -135,57 +165,79 @@ private void doRemoveDataNode(String[] args)
+              .filter(
+                  location ->
+                      endPoints.stream()
+                          .anyMatch(nodeCoordinate -> nodeCoordinate.matches(location)))
+              .collect(Collectors.toList());
+    } catch (TException | ClientManagerException e) {
+      LOGGER.error("Get data node locations failed", e);
+    }
+
+    return dataNodeLocations;
+  }
+
+  protected List<NodeCoordinate> parseCoordinates(String coordinatesString) {
+    // Multiple coordinates are separated by ","
+    String[] coordinateStrings = coordinatesString.split(",");
+    List<NodeCoordinate> coordinates = new ArrayList<>(coordinateStrings.length);
+    for (String coordinate : coordinateStrings) {
+      // If the string contains a ":" then this is an IP+Port configuration
+      if (coordinate.contains(":")) {
+        coordinates.add(new NodeCoordinateIP(UrlUtils.parseTEndPointIpv4AndIpv6Url(coordinate)));
       }
-    } catch (BadNodeUrlException e) {
-      try (ConfigNodeClient client =
-          ConfigNodeClientManager.getInstance().borrowClient(ConfigNodeInfo.CONFIG_REGION_ID)) {
-        for (String id : args.split(",")) {
-          if (!isNumeric(id)) {
-            LOGGER.warn("Incorrect id format {}, skipped...", id);
-            continue;
-          }
-          List<TDataNodeLocation> nodeLocationResult =
-              client
-                  .getDataNodeConfiguration(Integer.parseInt(id))
-                  .getDataNodeConfigurationMap()
-                  .values()
-                  .stream()
-                  .map(TDataNodeConfiguration::getLocation)
-                  .collect(Collectors.toList());
-          if (nodeLocationResult.isEmpty()) {
-            LOGGER.warn("DataNode {} is not in cluster, skipped...", id);
-            continue;
-          }
-          if (!dataNodeLocations.contains(nodeLocationResult.get(0))) {
-            dataNodeLocations.add(nodeLocationResult.get(0));
-          }
-        }
-      } catch (TException | ClientManagerException e1) {
-        LOGGER.error("Get data node locations failed", e);
+      // In the other case, we expect it to be a numeric value referring to the node-id
+      else if (NumberUtils.isCreatable(coordinate)) {
+        coordinates.add(new NodeCoordinateNodeId(Integer.parseInt(coordinate)));
       }
     }
-    return dataNodeLocations;
+    return coordinates;
+  }
+
+  protected interface NodeCoordinate {
+    // Returns true if the given location matches this coordinate
+    boolean matches(TDataNodeLocation location);
+  }
+
+  /** Implementation of a NodeCoordinate that uses ip and port to match. */
+  protected static class NodeCoordinateIP implements NodeCoordinate {
+    private final TEndPoint endPoint;
+
+    public NodeCoordinateIP(TEndPoint endPoint) {
+      this.endPoint = endPoint;
+    }
+
+    @Override
+    public boolean matches(TDataNodeLocation location) {
+      return endPoint.ip.equals(location.getClientRpcEndPoint().getIp())
```

Comment:
use internalRPCEndpoint?

Reply:
I think I didn't quite understand what the internal rpc endpoint is ... it this a second rpc endpoint which is used by the cluster internally? If yes, I would indeed prefer using that.

Reply:
Yes, each datanode has an external RPC and an internal RPC. The external RPC is meant to be easy to use, so it's possible to use 0.0.0.0 so that it doesn't have to worry about specific IP addresses, which makes it easier for users to use. However, 0.0.0.0 cannot be used for internal IP, so it can be used to identify nodes

Reply:
But 0.0.0.0 doesn't help if I want to connect to a node ... assuming I'm connected to one device (I obviously need to know one IP address in order to query it) ... I want to shut down that node in order to do something, I run the "give me the other data nodes data" and then I get a list of 0.0.0.0 ... so how would I connect to the next? 
I do understand that setting it up for 0.0.0.0 and that just means: Listen on all available IP addresses is simple for setup. But not for connecting to. So I would propose to use 0.0.0.0 for setup but internally to maintain a list of IPs in that case.

Reply:
![img_v3_02d3_8170a440-1d04-4e7d-83f3-5486e5deaafg](https://github.com/user-attachments/assets/7e97a74d-0b04-4b88-981b-f400c06ff0f9)
Above, the server is configured with three network cards
If 127.0.0.1 is configured, it can only be accessed locally
If one of these cards is configured, for example, 172.16.17.15, the network segment on which the other two cards are located cannot be accessed
If you configure 0.0.0.0, all will be accessible

If you want to take the ip of the communication between the cn and the dn, the actual ip is the internal_address rather than client_address. Then it makes less sense for us to separate the two ips.

Reply:
In any case, whether or not 0.0.0.0 is allowed to register and Remove datanodes is a separate matter. For the former, it feels like we could start a mailing list discussion to see if we need the next PR; Regarding the latter, we can fix this bug first?

Reply:
I used the client_address because it looked as if this was the one being used before (you can't configure 0.0.0.0 as internal_address I was told). 
All in all the IP actually doesn't really matter, as the values are simply used to filter items in the list of nodes returned by the config-node. If we use the Node-ID, that's absolutely fine. If we use anything else, it must be possible to identify a node ... using 0.0.0.0 didn't work as identifier, because all had the same value. 

If you ask me, I think it would be good to track the internal_address and a list of external_addresses. (As you are not guaranteed to be able to connect to the internal_address I think the list of external_addresses would be more important. So in above case, for example the internal_address could be 127.0.0.1, and if the external_address is set to 0.0.0.0, then it should list "11.101.17.0, 172.16.17.0, 172.17.0.0". If a user now used any of these external addresses in the command to remove the data-node, this would be clear which one to remove as only one data node will have the same ip-port combination.

Reply:
+1 for identifying nodes solely by nodeid. This way we don't have to worry about the function definition.

Let's wait for the community to react. If the majority is in favor, let's remove the IP identifier from this PR

Reply:
+1 to that ... would be the cleaner option.
(However for a user it would still be hard to spot the id of the node to remove as all have the same ip of 0.0.0.0)